File: subprojects/docs/src/snippets/plugins/buildServiceFromWorkAction/groovy/buildSrc/src/main/java/WebServer.java
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.services.BuildService;
import org.gradle.api.services.BuildServiceParameters;

import java.net.URI;
import java.net.URISyntaxException;

public abstract class WebServer implements BuildService<WebServer.Params>, AutoCloseable {

    // Some parameters for the web server
    interface Params extends BuildServiceParameters {
        Property<Integer> getPort();

        DirectoryProperty getResources();
    }

    private final URI uri;

    public WebServer() throws URISyntaxException {
        // Use the parameters
        int port = getParameters().getPort().get();
        uri = new URI(String.format("https://localhost:%d/", port));

        // Start the server ...

        System.out.println(String.format("Server is running at %s", uri));
    }

    // A public method for tasks to use
    public URI getUri() {
        return uri;
    }

    @Override
    public void close() {
        // Stop the server ...
    }
}

File: subprojects/docs/src/snippets/plugins/buildService/kotlin/buildSrc/src/main/kotlin/WebServer.kt
import org.gradle.api.services.BuildService
import org.gradle.api.services.BuildServiceParameters
import org.gradle.api.provider.Property
import org.gradle.api.file.DirectoryProperty
import java.net.URI

abstract class WebServer : BuildService<WebServer.Params>, AutoCloseable {

    // Some parameters for the web server
    interface Params : BuildServiceParameters {
        val port: Property<Int>
        val resources: DirectoryProperty
    }

    // A public property for tasks to use
    val uri: URI

    init {
        // Use the parameters
        val port = parameters.port.get()
        uri = URI("https://localhost:$port/")

        // Start the server ...

        println("Server is running at $uri")
    }

    override fun close() {
        // Stop the server ...
    }
}
